src/ray/gcs/gcs_server/gcs_placement_group_manager.cc

```diff
@@ -164,28 +164,31 @@ void GcsPlacementGroupManager::RegisterPlacementGroup(
     }
   }
 
-  // Mark the callback as pending and invoke it after the placement_group has been
-  // successfully created.
-  placement_group_to_register_callback_[placement_group->GetPlacementGroupID()] =
-      std::move(callback);
+  placement_group_to_register_callbacks_[placement_group->GetPlacementGroupID()]
+      .emplace_back(std::move(callback));
   registered_placement_groups_.emplace(placement_group->GetPlacementGroupID(),
                                        placement_group);
   pending_placement_groups_.emplace_back(placement_group);
 
   RAY_CHECK_OK(gcs_table_storage_->PlacementGroupTable().Put(
       placement_group_id, placement_group->GetPlacementGroupTableData(),
       [this, placement_group_id, placement_group](Status status) {
+        // The backend storage is supposed to be reliable, so the status must be ok.
         RAY_CHECK_OK(status);
         if (!registered_placement_groups_.contains(placement_group_id)) {
-          auto iter = placement_group_to_register_callback_.find(placement_group_id);
-          if (iter != placement_group_to_register_callback_.end()) {
-            std::stringstream stream;
-            stream << "Placement group of id " << placement_group_id
-                   << " has been removed before registration.";
-            iter->second(Status::NotFound(stream.str()));
-            placement_group_to_register_callback_.erase(iter);
-          }
+          // The placement group registration is synchronous, so if we found the placement
+          // group was deleted here, it must be triggered by a job crash,
+          // we will return directly in this case.
+          RAY_CHECK(placement_group_to_register_callbacks_.count(placement_group_id) ==
```

Comment:
If we found the placement group was deleted here, GCS will crash?

Reply:
hmm.. why GCS will crash?  the `register callback` will remove when `RemovePlacementGroup` was called.